happytransformer/qa/trainer.py

```diff
@@ -7,31 +7,80 @@
+from happytransformer.qa.default_args import ARGS_QA_TRAIN, ARGS_QA_EVAl, ARGS_QA_TEST
+
+@dataclass
+class QATrainArgs:
+    learning_rate: float = ARGS_QA_TRAIN["learning_rate"]
+    num_train_epochs: int = ARGS_QA_TRAIN["num_train_epochs"]
+    weight_decay: float = ARGS_QA_TRAIN["weight_decay"]
+    adam_beta1: float = ARGS_QA_TRAIN["adam_beta1"]
+    adam_beta2: float = ARGS_QA_TRAIN["adam_beta2"]
+    adam_epsilon: float = ARGS_QA_TRAIN["adam_epsilon"]
+    max_grad_norm:  float = ARGS_QA_TRAIN["max_grad_norm"]
+    save_preprocessed_data: bool = ARGS_QA_TRAIN["save_preprocessed_data"]
+    save_preprocessed_data_path: str = ARGS_QA_TRAIN["save_preprocessed_data_path"]
+    load_preprocessed_data: bool = ARGS_QA_TRAIN["load_preprocessed_data"]
+    load_preprocessed_data_path: str = ARGS_QA_TRAIN["load_preprocessed_data_path"]
+
+
+@dataclass
+class QAEvalArgs:
+    save_preprocessed_data: bool = ARGS_QA_EVAl["save_preprocessed_data"]
+    save_preprocessed_data_path: str = ARGS_QA_EVAl["save_preprocessed_data_path"]
+    load_preprocessed_data: bool = ARGS_QA_EVAl["load_preprocessed_data"]
+    load_preprocessed_data_path: str = ARGS_QA_EVAl["load_preprocessed_data_path"]
+
+@dataclass
+class QATestArgs:
+    save_preprocessed_data: bool = ARGS_QA_TEST["save_preprocessed_data"]
+    save_preprocessed_data_path: str = ARGS_QA_TEST["save_preprocessed_data_path"]
+    load_preprocessed_data: bool = ARGS_QA_TEST["load_preprocessed_data"]
+    load_preprocessed_data_path: str = ARGS_QA_TEST["load_preprocessed_data_path"]
 
 class QATrainer(HappyTrainer):
     """
     Trainer class for HappyTextClassification
     """
 
-    def train(self, input_filepath, args):
+    def train(self, input_filepath, dataclass_args: QATrainArgs):
         """
         See docstring in HappyQuestionAnswering.train()
         """
-        #todo: add time elapsed and test time remaining similar to what is within eval
-
-        contexts, questions, answers = self._get_data(input_filepath)
-        self.__add_end_idx(contexts, answers)
-        encodings = self.tokenizer(contexts, questions, truncation=True, padding=True)
-        self.__add_token_positions(encodings, answers)
-        dataset = QuestionAnsweringDataset(encodings)
-        self._run_train(dataset, args)
-
-
-    def eval(self, input_filepath):
+        if not dataclass_args.load_preprocessed_data:
+            self.logger.info("Preprocessing dataset...")
+            contexts, questions, answers = self._get_data(input_filepath)
+            self.__add_end_idx(contexts, answers)
+            encodings = self.tokenizer(contexts, questions, truncation=True, padding=True)
+            self.__add_token_positions(encodings, answers)
+            print("encodings", encodings)
+        else:
+            self.logger.info("Loading dataset from %s...", dataclass_args.load_preprocessed_data_path)
+            encodings = self._get_preprocessed_data(dataclass_args.load_preprocessed_data_path)
```

Comment:
Could these two blocks be switched? `(if not) / else` is a confusing structure.

Reply:
Yes, I just wanted to put the common/default case first.